### PR TITLE
feat: Add compact protobuf format

### PR DESCRIPTION
**Pull request by mkg20001**
This adds a significantly smaller (about 30%) protobuf format
It's smaller because `.toJSON()` encodes the binary keys as base64 which makes it bigger and is sometimes not needed - like when transferring the id over the network
Additionally an `excludePriv` option is added because the removal of the private key is not as easy as in json

### CODE REVIEW

**Comment by diasdavid** [comment]
Why not use CBOR?

**Comment by mkg20001** [comment]
Hmmm... which one is smaller?

File: src/index.js [hunk 1]
```diff
@@ -9,6 +9,9 @@ const crypto = require('libp2p-crypto')
 const assert = require('assert')
 const waterfall = require('async/waterfall')
 
+const protons = require('protons')
+const {PeerIdProto} = protons('message PeerIdProto { required bytes id = 1; bytes pubKey = 2; bytes privKey = 3; }')
```
Comment (pgte) [inline src/index.js]: Perhaps add this to a separate file? Looks weird here.

Comment (mkg20001) [reply src/index.js]: I think I should, was just lazy because it's so small ;)

**Comment by mkg20001** [comment]
@pgte Requested changes applied. Could you please re-review?

File: src/index.js [hunk 2]
```diff
@@ -221,6 +231,62 @@ exports.createFromPrivKey = function (key, callback) {
   })
 }
 
+exports.createFromProtobuf = function (buf, callback) {
```
Comment (pgte) [inline src/index.js]: Could you just somehow use .createFromJSON so that we don't have the same logic twice?

**Comment by pgte** [comment]
Also, Travis CI is complaining..

File: src/index.js (hunk 2 above)
Comment (mkg20001) [reply src/index.js]: .createFromJSON expects the data to be in base64. I could make a common function out of the validation logic.

Comment (pgte) [reply src/index.js]: @mkg20001 that would be good.

**Comment by mkg20001** [comment]
@pgte Applied requested changes, additionally fixed some other things

**Comment by pgte** [comment]
@mkg20001 great, thank you!
CI is complaining, could you check that?

**Comment by mkg20001** [comment]
@pgte Yes, although it's unclear to me what exactly is wrong with the commit message. It only shows it's not in the required format, not where exactly the error is

**Comment by pgte** [comment]
@VictorBjelkholm could you help clarifying this question about the commit message?

**Comment by mkg20001** [comment]
Updated this PR to async/await

File: src/index.js [hunk 3]
```diff
@@ -8,6 +8,7 @@ const mh = require('multihashes')
 const cryptoKeys = require('libp2p-crypto/src/keys')
 const assert = require('assert')
 const withIs = require('class-is')
+const {PeerIdProto} = require('./proto')
```
Comment (jacobheun) [inline src/index.js]:
```suggestion
const { PeerIdProto } = require('./proto')
```

File: src/index.js [hunk 4]
```diff
@@ -232,6 +242,49 @@ exports.createFromJSON = async (obj) => {
   return new PeerIdWithIs(id, privKey, pub)
 }
 
+exports.createFromProtobuf = async (buf) => {
+  if (typeof buf === 'string') {
+    buf = Buffer.from(buf, 'hex')
+  }
+
+  let {id, privKey, pubKey} = PeerIdProto.decode(buf)
```
Comment (jacobheun) [inline src/index.js]:
```suggestion
  let { id, privKey, pubKey } = PeerIdProto.decode(buf)
```

**Comment by mkg20001** [comment]
@jacobheun Thx, changes commited

File: test/peer-id.spec.js [hunk 5]
```diff
@@ -86,6 +86,15 @@ describe('PeerId', () => {
...
   })
 
+  it('Recreate from Protobuf', async () => {
+    const id = await PeerId.createFromProtobuf(testId.marshaled)
+    expect(testIdB58String).to.equal(id.toB58String())
+    const encoded = Buffer.from(testId.privKey, 'base64')
+    const id2 = await PeerId.createFromPrivKey(encoded)
+    expect(testIdB58String).to.equal(id2.toB58String())
+    expect(id.marshalPubKey()).to.deep.equal(id2.marshalPubKey())
+  })
```
Comment (jacobheun) [inline test/peer-id.spec.js]: @mkg20001 it looks like `marshal` isn't being tested, can we validate that marshaling the peerid gets us the same `testId.marshaled` value?

Other than that, this looks good.

Comment (mkg20001) [reply test/peer-id.spec.js]: Done

**Comment by jacobheun** [comment]
released in 0.13.2

**Comment by mkg20001** [comment]
Nice :tada:

**Pull request merged**